crypto/keys/sr25519/sr25519.go

```diff
@@ -0,0 +1,151 @@
+	cryptotypes "github.com/cosmos/cosmos-sdk/crypto/types"
+	"github.com/cosmos/cosmos-sdk/types/errors"
+)
+
+var _ cryptotypes.PrivKey = &PrivKey{}
+var _ codec.AminoMarshaler = &PrivKey{}
+
+const (
+	PubKeySize  = 32
+	PrivKeySize = 32
+	keyType     = "sr25519"
+	PrivKeyName = "tendermint/PrivKeySr25519"
+	PubKeyName  = "tendermint/PubKeySr25519"
+)
+
+// Bytes returns the byte representation of the Private Key.
+func (privKey *PrivKey) Bytes() []byte {
+	return privKey.MiniSecretKey
+}
+
+// Sign produces a signature on the provided message.
+func (privKey *PrivKey) Sign(msg []byte) ([]byte, error) {
+	return tmsr25519.PrivKey(privKey.MiniSecretKey).Sign(msg)
+}
+
+// PubKey gets the corresponding public key from the private key.
+func (privKey *PrivKey) PubKey() cryptotypes.PubKey {
+	return &PubKey{Key: tmsr25519.PrivKey(privKey.MiniSecretKey).PubKey().Bytes()}
+}
+
+// Equals - you probably don't need to use this.
+// Runs in constant time based on length of the
+func (privKey *PrivKey) Equals(other cryptotypes.LedgerPrivKey) bool {
+	return privKey.Type() == other.Type() && subtle.ConstantTimeCompare(privKey.Bytes(), other.Bytes()) == 1
+}
+
+func (privKey *PrivKey) Type() string {
+	return keyType
+}
+
+// MarshalAmino overrides Amino binary marshalling.
+func (privKey PrivKey) MarshalAmino() ([]byte, error) {
+	return privKey.MiniSecretKey, nil
+}
+
+// UnmarshalAmino overrides Amino binary marshalling.
+func (privKey *PrivKey) UnmarshalAmino(bz []byte) error {
+	if len(bz) != PrivKeySize {
+		return fmt.Errorf("invalid privkey size")
+	}
+	privKey.MiniSecretKey = bz
+
+	return nil
+}
+
+// MarshalAminoJSON overrides Amino JSON marshalling.
+func (privKey PrivKey) MarshalAminoJSON() ([]byte, error) {
+	// When we marshal to Amino JSON, we don't marshal the "key" field itself,
+	// just its contents (i.e. the key bytes).
+	return privKey.MarshalAmino()
+}
+
+// UnmarshalAminoJSON overrides Amino JSON marshalling.
+func (privKey *PrivKey) UnmarshalAminoJSON(bz []byte) error {
+	return privKey.UnmarshalAmino(bz)
+}
+
+// GenPrivKey generates a new sr25519 private key.
+// It uses OS randomness in conjunction with the current global random seed
+// in tendermint/libs/common to generate the private key.
+func GenPrivKey() *PrivKey {
+	return &PrivKey{MiniSecretKey: tmsr25519.GenPrivKey().Bytes()}
+}
+
+// GenPrivKeyFromSecret hashes the secret with SHA2, and uses
+// that 32 byte output to create the private key.
+// NOTE: secret should be the output of a KDF like bcrypt,
+// if it's derived from user input.
+func GenPrivKeyFromSecret(secret []byte) *PrivKey {
+	return &PrivKey{MiniSecretKey: tmsr25519.GenPrivKeyFromSecret(secret).Bytes()}
+}
+
+//-------------------------------------
+
+var _ cryptotypes.PubKey = &PubKey{}
+var _ codec.AminoMarshaler = &PubKey{}
+
+// Address returns a Bitcoin style addresses: RIPEMD160(SHA256(pubkey))
+func (pubKey *PubKey) Address() crypto.Address {
+	return tmsr25519.PubKey(pubKey.Key).Address()
```

Comment:
This has to be ADR-40 compatible. You can use this code: https://github.com/cosmos/cosmos-sdk/blob/master/crypto/keys/secp256r1/pubkey.go#L32
(you will need to copy part of the `m.Key` function).

Reply:
You mean ADR 028?

Reply:
Yes